### PR TITLE
Support index purge in dreyfus.

### PR DESCRIPTION
**Pull request by chguocloudant**
- use couch_db:fold_purged_docs to get list of documents to be purged
- update and delete index in clouseau for purged document
- write test logic to cover changed code

### CODE REVIEW

File: src/dreyfus_index.erl
Comment (mayya-sharipova) [inline src/dreyfus_index.erl:229]: it may be better to have a single message `get_update_purge_seqs` or smth, rather than to send two messages

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [inline src/dreyfus_index_updater.erl]: I would make lines 35-65 as a separate function, that either returns `{ok, NewPurgeSeq}` or `reset`.

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [inline src/dreyfus_index_updater.erl]: you don't need to do `set_purge_seq` in the loop. Similarly to update_seq you can commit it periodically (e.g. every minute) or at the end.

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [inline src/dreyfus_index_updater.erl]: Instead of list for Acc you can use just a starting purge_seq, so at the end you will get your final purge_seq for the accumulator, and you don't need to do `lists:max(PSeqs)` on line 56.

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [inline src/dreyfus_index_updater.erl]: This will deleted all search indexes for the database, we need just delete the current index.

File: src/dreyfus_index.erl
Comment (mayya-sharipova) [inline src/dreyfus_index.erl]: we need to discuss in the #cloudant-search channel how to properly reset the index. I think if we are resetting the index, we need to kill the existing dreyfus_index process, corresponding clouseau_process, remove all corresponding lucene files, and after that initiate new clouseau & dreyfus processes. And this what dreyfus_index_manager should do

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: I think the purge seq will be loss if an exception happened and the loop is not finished yet.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: You mean the we can get final_purge_seq by adding the count of `PurgeSeqIdRevs` and starting purge_seq?
is the `PurgeSeqIdRevs` consecutive?

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [reply src/dreyfus_index_updater.erl]: @chguocloudant  even with `clouseau_rpc:set_purge_seq` you are not committing it to disk, so if the process crashes, you will still loose a new `purge_seq`

File: src/dreyfus_index_updater.erl
Comment (mayya-sharipova) [reply src/dreyfus_index_updater.erl]: I mean this:
```
NewPurgeSeq = lists:foldr(fun({PSeq, Id, _}, _Acc) ->
                    clouseau_rpc:delete(IndexPid, Id),
                    clouseau_rpc:set_purge_seq(IndexPid, PSeq),
                    PSeq
                end, IdxPurgeSeq, PurgeSeqIdRevs),
```

Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: got it. 
And I will check if the `PurgeSeqIdRevs` is already sorted by PurgeSeq.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: got it, thank you.

File: src/dreyfus_index.erl
Comment (rnewson) [inline src/dreyfus_index.erl]: why did this section change?

File: src/dreyfus_index.erl
Comment (rnewson) [inline src/dreyfus_index.erl:171]: why did this section change?

File: src/dreyfus_index.erl
Comment (rnewson) [inline src/dreyfus_index.erl]: why was this function extracted?

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: why did this argument disappear?

**Review by rnewson — COMMENTED** [review]
This PR mixes the purge changes along with stylistic changes and refactorings. Please remove everything except what is needed to support index purge.

File: src/dreyfus_index.erl
Comment (chguocloudant) [reply src/dreyfus_index.erl]: There are some changes here before, but they are removed.
I will change it back to origin status.

File: src/dreyfus_index.erl
Comment (chguocloudant) [reply src/dreyfus_index.erl:171]: these lines are extracted to a new function to make it easier to understand.
I followed the comments in hastings from Paul Davis. But the new function here is only called once.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: We used a new function `couch_db:fold_changes` here, and this function need a function with 2 arguments.

File: src/dreyfus_index.erl
Comment (chguocloudant) [reply src/dreyfus_index.erl:171]: I have modified it and commit it later.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: I don't think this is the correct behaviour. The purge API is on id/rev pairs, the request might leave some revisions alone, in which case the index should reflect that view. This code appears to remove the document from the index in all cases.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [reply src/dreyfus_index_updater.erl]: Why did the call to enum_docs_since change to a call to fold_changes?

File: src/dreyfus_rpc.erl
Comment (rnewson) [inline src/dreyfus_rpc.erl]: Please remove the extraction of get_index_pid from this PR so we can focus on the purge feature change.

File: test/dreyfus_purge_test.erl
Comment (rnewson) [inline test/dreyfus_purge_test.erl:26]: It would help if you added a test where only some revisions of a document are purged, we can then assert that the remaining revision is indexed correctly.

**Review by rnewson — CHANGES_REQUESTED** [review]

File: src/dreyfus_rpc.erl
Comment (chguocloudant) [reply src/dreyfus_rpc.erl]: OK, I will remove this function.

File: test/dreyfus_purge_test.erl
Comment (chguocloudant) [reply test/dreyfus_purge_test.erl:26]: OK, I will make a new function to test this case.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: This change is used for pluggable storage engine API. The purge issue is based on plugable storage engine API.
But I think it should be commit separately.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: This `FoldFun` is enum the purged docs, and delete the invalid index. But the index will be rebuilt later.
The purge API is work on id/rev pairs, but one doc has only one index. So if the doc is changed, the index should be delete first and then rebuild it.

**Comment by rnewson** [comment]
I'm out on vacation until Jan 3rd but can comment today. 

That's not how we want to update the index. It should go from its current state to the new state without an intermediate state where the document is deleted from the index.

**Comment by chguocloudant** [comment]
Hi Robert, good day.
 
I think I should express our steps in purging index.This is our ways to purge and rebuild index as below.
 
If a doc has only one leaf node. This doc is purged and also be deleted.Search this doc, and trigged the function `dreyfus_index_updater:update`.Firstly, fold purged docs by `couch_db:fold_purged_docs/5`. And the doc's index will be deleted in the fold function.Secondly, fold updated docs by `couch_db:fold_changes` and the fold function is `load_docs/2`. The doc is deleted already, so do nothing about this doc in the fold function.
 
If a doc has more then one leaf node. One of the leaf node is purged but this doc is not deleted.Search this doc, and trigged the function `dreyfus_index_updater:update`.Firstly, fold purged docs by `couch_db:fold_purged_docs/5`. And the doc's index will be deleted in the fold function.Secondly, fold updated docs by `couch_db:fold_changes` and the fold function is `load_docs/2`. The doc exist but there is no index, so the index will be updated(rebuilt).
 
I think it achieves the result we want too.
 
I understand your meaning is that the doc should goes a)from original state to deleted state or b)from original state to updated state in fold function when we fold the purge seq. Just like the operations in `load_docs/2`.a) the doc is purged and deleted.b) the doc is purged but not deleted.And current code goes from original state to deleted state first and then goes to updated state if the doc still exist.
 
But we can only get `Id` and purged `Rev` in the fold function, so we don't have enough info to update(rebuild) index directly here.
 
What do you think?
 
Chang Hua Guo (郭昌华)
E-mail: chguo@cn.ibm.com
 
 
----- Original message -----From: Robert Newson <notifications@github.com>To: cloudant-labs/dreyfus <dreyfus@noreply.github.com>Cc: Chang Hua DG Guo/China/IBM@IBMCN, Mention <mention@noreply.github.com>Subject: Re: [cloudant-labs/dreyfus] Support index purge in dreyfus. (#11)Date: Fri, Dec 23, 2016 4:58 PM I'm out on vacation until Jan 3rd but can comment today.That's not how we want to update the index. It should go from its current state to the new state without an intermediate state where the document is deleted from the index.Sent from my iPhone> On 23 Dec 2016, at 07:47, Chang Hua Guo <notifications@github.com> wrote:>> @chguocloudant commented on this pull request.>> In src/dreyfus_index_updater.erl:>> > @@ -31,6 +31,14 @@ update(IndexPid, Index) ->> erlang:put(io_priority, {view_update, DbName, IndexName}),> {ok, Db} = couch_db:open_int(DbName, []),> try> + {ok, IdxPurgeSeq} = clouseau_rpc:get_purge_seq(IndexPid),> + FoldFun = fun(PurgeSeq, {Id, _Revs}, Acc) ->> + clouseau_rpc:delete(IndexPid, Id),> This FoldFun is enum the purged docs, and delete the invalid index. But the index will be rebuilt later.> The purge API is work on id/rev pairs, but one doc has only one index. So if the doc is changed, the index should be delete first and then rebuild it.>> —> You are receiving this because you commented.> Reply to this email directly, view it on GitHub, or mute the thread.>

File: src/dreyfus_index_updater.erl
Comment (rnewson) [reply src/dreyfus_index_updater.erl]: ok, we can't merge this PR until the pluggable storage engine API is also merged.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [reply src/dreyfus_index_updater.erl]: Hm, I'm still not convinced. If we look at couch_index/couch_mrview that seems to do the right thing. It purges the items and then updates the index, before returning the updated mrview state to the caller. Here we will make the intermediate state, where the Documents in the Lucene index are removed, visible to search results.

**Comment by chguocloudant** [comment]
Hi @rnewson . 
I am not sure if I express myself clear enough. If you have any question please leave some comments.
Thank you very much.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: I read couch_index/couch_mrview code and I think it is same to drefus. It is also delete first and then rebuild it.
If we purged a document and then search doc. Here is the steps in dreyfus.
1. fold purged docs and it will delete the indexes of purged docs.
2. fold db changes and it will rebuild the indexes of docs if they are still exist.
3. do search.

The operations are triggered by search request. Delete and update index first and then search and return the result to the caller.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl:88]: Proc needs to be returned when we're done (a try/after is usual).

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: `[{Id, Rev} | Acc]` is more idiomatic (and faster). If order matters, then reverse the list at the end.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: just `Acc;` here. Rather than assign a variable in every path, it's more idiomatic to use the result of the case statement (i.e, `Acc0 = case couch_db...`).

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: this section is a cut-and-paste from elsewhere in the file, please extract it as a function that is called from both places instead.

File: src/dreyfus_index_updater.erl
Comment (rnewson) [inline src/dreyfus_index_updater.erl]: this does nothing as the result is discarded.

**Review by rnewson — CHANGES_REQUESTED** [review]

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl:88]: Got it. I will modify it and commit it later.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: Got it, will modify and commit it later.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: OK, I will extract it as a new function.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: Sorry, I will remove it.

File: src/dreyfus_index_updater.erl
Comment (chguocloudant) [reply src/dreyfus_index_updater.erl]: Got it. will modify it.

File: test/dreyfus_purge_test.erl
Comment (rnewson) [inline test/dreyfus_purge_test.erl:60]: this could be improved with `?assertMatch`, i.e;

`?assertMatch({ok, _, 5, _, _, _}, dreyfus_search(DbName, Query)),`

File: test/dreyfus_purge_test.erl
Comment (rnewson) [inline test/dreyfus_purge_test.erl:71]: same comment as above

**Review by rnewson — APPROVED** [review]
I'd like to see the eunit tests improved as indicated but I think this work is complete, thanks!

A reminder that we can't merge this until the pluggable storage API is merged into couchdb master.

Also, we might have a concern with the idrev list we build during the fold which helps us not make duplicate updates to the search index. If that was a very large list, it might be a problem. What do you think?

File: test/dreyfus_purge_test.erl
Comment (chguocloudant) [reply test/dreyfus_purge_test.erl:60]: Thanks, got it, I will modify it.

File: test/dreyfus_purge_test.erl
Comment (chguocloudant) [reply test/dreyfus_purge_test.erl:71]: Thanks, got it, I will modify it.

**Comment by chguocloudant** [comment]
Thanks. I won't merge it until the pluggable storage API is merged into couchdb master.
The size of the IdRev list depends on how many docs user purged in a period. I think the size won't be too large generally.
But if this happened, I don't think what can be done in dreyfus to fix this problem. We can try to use some more effective functions to shorten execution time. 
I think we can set a threshold to trigger the index purging automatically. It maybe can avoid this problem.

**Pull request closed without merging**
